File: data_structure/02_stack/stack.c
#include <stdio.h>
#include <stdbool.h>

#define INITIAL_CAPACITY 10

struct Stack {
    int* arr;
    int top;
    int capacity;
};

void initialize(struct Stack* stack) {
    stack->capacity = INITIAL_CAPACITY;
    stack->arr = (int*)malloc(sizeof(int) * stack->capacity);
    stack->top = -1;
}

void destroy(struct Stack* stack) {
    free(stack->arr);
}

bool isEmpty(struct Stack* stack) {
    return stack->top == -1;
}

bool isFull(struct Stack* stack) {
    return stack->top == stack->capacity - 1;
}

void expandStack(struct Stack* stack) {
    stack->capacity *= 2;
    stack->arr = (int*)realloc(stack->arr, sizeof(int) * stack->capacity);
}

void push(struct Stack* stack, int data) {
    if (isFull(stack)) {
        expandStack(stack);
    }
    stack->arr[++stack->top] = data;
}

int pop(struct Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack Underflow\n");
    }
    return stack->arr[stack->top--];
}

int peek(struct Stack* stack) {
    if (isEmpty(stack)) {
        printf("Stack is empty\n");
        return -1; 
    }
    return stack->arr[stack->top];
}

void clear(struct Stack* stack) {
    stack->top = -1;
}

int main() {
    struct Stack stack;
    initialize(&stack);

    push(&stack, 1);
    push(&stack, 2);
    push(&stack, 3);

    printf("%d\n", pop(&stack)); 
    printf("%d\n", pop(&stack));  
    printf("%d\n", peek(&stack)); 
    clear(&stack); 
    printf("%d\n", pop(&stack));

    destroy(&stack); 

    return 0;
}

File: data_structure/02_stack/go/stack.go
package main

import (
	"fmt"
)

type Stack struct {
	arr []int
}

func (s *Stack) initialize() {
	// Initialize an empty slice
	s.arr = []int{}
}

func (s *Stack) isEmpty() bool {
	return len(s.arr) == 0
}

func (s *Stack) push(data int) {
	s.arr = append(s.arr, data)
}

func (s *Stack) pop() int {
	if s.isEmpty() {
		fmt.Println("Stack Underflow")
		return -1
	}
	// Get the top element of the stack
	topElement := s.arr[len(s.arr)-1]
	// Remove the top element from the stack
	s.arr = s.arr[:len(s.arr)-1]
	return topElement
}

func (s *Stack) peek() int {
	if s.isEmpty() {
		fmt.Println("Stack is empty")
		return -1
	}
	return s.arr[len(s.arr)-1]
}

func (s *Stack) clear() {
	// Clear the stack, simply by creating a new empty slice
	s.arr = []int{}
}

func main() {
	var stack Stack
	stack.initialize()

	stack.push(1)
	stack.push(2)
	stack.push(3)

	fmt.Println(stack.pop())  // Output: 3
	fmt.Println(stack.pop())  // Output: 2
	fmt.Println(stack.peek()) // Output: 1 (peek without removing)
	stack.clear()             // Clear the stack
	fmt.Println(stack.pop())  // Output: Stack Underflow
}
